### ISSUE TITLE
[BUG] Trying to debug a test in VS Code with linetrace=True raises "TypeError: 'NoneType' object is not callable"

### ISSUE THREAD

**Issue by segevfiner** [cython/cython#4443]
**Describe the bug**
When trying to debug a pytest test in VS Code that imports a Cython module that has `linetrace` enabled, I'm getting a "TypeError: 'NoneType' object is not callable" at import time.

```
=================================== ERRORS ====================================
________________________ ERROR collecting test_spam.py ________________________
test_spam.py:1: in <module>
    import spam
spam.pyx:4: in init spam
    def f():
spam.pyx:4: in __Pyx_PyMODINIT_FUNC PyInit_spam(void)
    def f():
E   TypeError: 'NoneType' object is not callable
```

**To Reproduce**
Code to reproduce the behaviour:
`spam.pyx`:
```cython
# cython: language_level=3str, linetrace=True
# distutils: define_macros=CYTHON_TRACE_NOGIL=1

def f():
    return 1234
```

`test_spam.py`:
```python
import spam


def test_f():
    assert spam.f() == 1234
```

Build `spam.pyx` with `cythonize -i spam.pyx`

Then try and debug `test_f` with VS Code using the Python extension's Pytest support.

**Expected behavior**

**Environment (please complete the following information):**
 - OS: Windows
 - Python version 3.10.0
 - Cython version 0.29.24

**Comment by scoder** [comment]
Would be nice to support, but I'm not surprised that Python debugging has issues. I don't think anyone is seriously doing that. I suggest either a) using a C level debugger instead or b) debugging the uncompiled Python (.py) code and not writing  .pyx files (not sure if that's an option for you when wrapping `libpcap` & friends).

`gdb` has support for debugging Python, Cython and C code. Don't know if there is anything similar for Windows/MSVC.

**Comment by scoder** [comment]
(Not sure what a good label is – line tracing is (most?) commonly used for coverage reporting, so I'll put it there.)

**Comment by gentlegiantJGC** [comment]
I am having the same issue with cython 3.0.0a9 and PyCharm.
It would be nice to step through the code to debug issues if that is even possible. Not sure if that would require work from the PyCharm team as well or just on your end.
Coverage reporting would also be good as well.
The code runs just fine when run normally.

```
  File "C:/Users/james_000/AppData/Roaming/JetBrains/PyCharm2021.2/scratches/debug.py", line 1, in <module>
    import amulet_nbt
  File "D:\Data\GitHub\Amulet-NBT\amulet_nbt\__init__.py", line 1, in <module>
    from .value import BaseTag, BaseImmutableTag, BaseMutableTag, BaseValueType
  File "amulet_nbt\\value.pyx", line 1, in init amulet_nbt.value
  File "amulet_nbt\\util.pyx", line 1, in init amulet_nbt.util
  File "stringsource", line 1, in init amulet_nbt.errors
  File "amulet_nbt\\errors.pyx", line 1, in __Pyx_PyMODINIT_FUNC PyInit_errors(void)
TypeError: 'NoneType' object is not callable
```

**Comment by scoder** [comment]
Yeah, it's probably something that they should investigate on their side first. If only to find out what we should do better on ours. It's not unlikely that they would at least have to do … something … in the end. Or maybe *could* even do something on their side already.

In any case, it would appear easier for them to debug the problem than for us.

**Comment by segevfiner** [comment]
It looks like Cython is handling `f_trace` wrong. This is an internal member used by `trace_trampoline` to support a Python trace function that returns a different trace function than itself, basically the thread local trace function. Cython is setting it to `None`, but it should (Surprisingly and unlike other `PyObject*`) be `NULL`. Actually, Cython should just avoid touching the it at all, as it is only used by `trace_trampoline` internally for it's own operation and isn't supposed to be touched by code externally calling the `c_tracefunc` (Which again is a Python implementation detail, FUN!).

**Comment by scoder** [comment]
According to a comment in the code, it does that in order to enable line number reporting:
https://github.com/cython/cython/blob/f848bdf69717e68b919ef7a43b651f77bf4fb980/Cython/Utility/Profile.c#L292-L296

**Comment by segevfiner** [comment]
Yeah. But I'm not sure why... I don't see any reference to that effect in CPython...

**Comment by scoder** [comment]
https://bugs.python.org/issue42823

**Comment by QLutz** [comment]
Hi and sorry for reviving this thread a year and half later.

I'm having trouble understanding the underlying issue and related PR. Is there any hope that this could be fixed in the near future?

Thanks in advance !
